File: leetcode/1001-1999/1475-final-prices-with-a-special-discount-in-a-shop/FinalPricesWithASpecialDiscountInAShop.java
import java.util.Arrays;

public class FinalPricesWithASpecialDiscountInAShop {
  public static void main(String[] args) {
    int[] prices = { 10, 1, 1, 6 };

    System.out.println(Arrays.toString((new Solution()).finalPrices(prices)));
  }
}

class Solution {
  public int[] finalPrices(int[] prices) {
    for (int i = 0; i < prices.length; i++) {
      int j = i + 1;

      while (j < prices.length && prices[j] > prices[i]) {
        j++;
      }

      if (j < prices.length) {
        prices[i] -= prices[j];
      }
    }

    return prices;
  }
}

File: leetcode/1001-1999/1475-final-prices-with-a-special-discount-in-a-shop/FinalPricesWithASpecialDiscountInAShop.ts
function finalPrices(prices: number[]): number[] {
  const stack: number[] = [];

  prices.forEach(
    (value, i) => {
      while (stack.length > 0 && prices[stack[stack.length - 1]] >= value) {
        prices[stack.pop() ?? 0] -= value;
      }

      stack.push(i);
    });


  return prices;
};

const prices = [8, 4, 6, 2, 3];

console.log(finalPrices(prices));
